File: fastify_server/src/addons/route.addon.ts
import { pathToFileURL } from "node:url";
import type { Addon } from "../registry/registry.js";
import { createLoaderReport } from "../contract/index.js";
import { createLoaderLogger } from "../registry/loader_logger.js";
import { discoverFiles } from "./_discover.js";

const ROUTE_SUFFIXES = [".routes.mjs", ".routes.js"] as const;

export const routeAddon: Addon = {
  name: "route",
  priority: 30,
  async run(server, config, ctx) {
    const report = createLoaderReport("route");
    const log = createLoaderLogger("route", ctx.logger, report);
    const adapter = ctx.getAdapter();

    for (const dir of config.paths.routes) {
      let result;
      try {
        result = await discoverFiles(dir, ROUTE_SUFFIXES);
      } catch (err) {
        log.failed("discover", dir, err);
        continue;
      }
      log.scanDir(dir, result.matched.length, result.ignored.length);
      for (const p of result.ignored) log.ignored(p);
      report.discovered += result.matched.length;

      for (const file of result.matched) {
        let mod: { default?: unknown; mount?: unknown };
        try {
          mod = (await import(pathToFileURL(file).href)) as { default?: unknown; mount?: unknown };
          report.imported += 1;
          log.loaded(file);
        } catch (err) {
          log.failed("import", file, err);
          continue;
        }
        const fn = (mod.default ?? mod.mount) as
          | ((s: unknown, c: unknown) => void | Promise<void>)
          | undefined;
        if (typeof fn !== "function") {
          report.skipped += 1;
          log.failed("shape", file, "module has no default or mount function");
          continue;
        }
        try {
          await adapter.registerRoutes(server, fn as never);
          report.registered += 1;
          log.registered(file);
        } catch (err) {
          log.failed("register", file, err);
        }
      }
    }

    return report;
  },
};

export default routeAddon;

File: fastapi_server/fastapi_server/bootstrap/addons/route_addon.py
from __future__ import annotations

from fastapi_server.bootstrap.contract import create_loader_report
from fastapi_server.bootstrap.registry.registry import Addon
from fastapi_server.bootstrap.registry.loader_logger import create_loader_logger
from fastapi_server.bootstrap.addons._discover import discover_files, import_file

_SUFFIXES = (".routes.py",)


def _run(server, config, ctx):
    report = create_loader_report("route")
    log = create_loader_logger("route", ctx.logger, report)
    adapter = ctx.get_adapter()

    for d in config.paths.routes:
        try:
            result = discover_files(d, _SUFFIXES)
        except Exception as err:  # noqa: BLE001
            log.failed("discover", d, err)
            continue
        log.scan_dir(d, len(result.matched), len(result.ignored))
        for p in result.ignored:
            log.ignored(p)
        report.discovered += len(result.matched)

        for file in result.matched:
            try:
                mod = import_file(file, module_name_prefix="polyglot_routes")
                report.imported += 1
                log.loaded(file)
            except Exception as err:  # noqa: BLE001
                log.failed("import", file, err)
                continue

            fn = getattr(mod, "mount", None) or getattr(mod, "router", None) or getattr(mod, "default", None)
            if fn is None:
                report.skipped += 1
                log.failed("shape", file, "module exposes no 'mount', 'router' or 'default'")
                continue

            try:
                if callable(fn):
                    adapter.register_routes(server, fn)
                else:
                    # Assume it's already an APIRouter-like object
                    adapter.register_routes(server, lambda _s, _c, _fn=fn: _fn)
                report.registered += 1
                log.registered(file)
            except Exception as err:  # noqa: BLE001
                log.failed("register", file, err)

    return report


route_addon = Addon(name="route", priority=30, run=_run)
